sync/sync.go

```diff
@@ -0,0 +1,345 @@
+		}
+		ch <- data.Block
+	}
+
+	return ch, s.p.SendAsyncRequest(BLOCK, payload, peer, foo)
+}
+
+func (s *Syncer) SendLayerHashRequest(peer Peer, layer uint32) ([]byte, error) {
+	log.Debug("send Layer hash request Peer: ", "layer: ", layer)
+
+	data := &pb.LayerHashReq{Layer: layer}
+	payload, err := proto.Marshal(data)
+	if err != nil {
+		log.Error("could not marshall layer hash request")
+		return nil, err
+	}
+
+	msg, err := s.p.SendRequest(LAYER_HASH, payload, peer, s.config.requestTimeout)
+	if err != nil {
+		log.Error("could not send layer hash request ", err)
+		return nil, err
+	}
+
+	res := &pb.LayerHashResp{}
+	err = proto.Unmarshal(msg.([]byte), res)
+	if err != nil {
+		log.Error("could not unmarshal layer hash response ", err)
+		return nil, err
+	}
+
+	return res.Hash, nil
+}
+
+func (s *Syncer) SendLayerIDsRequest(peer Peer, idx uint32) (chan []uint32, error) {
+
+	data := &pb.LayerIdsReq{Layer: idx}
+	payload, err := proto.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	ch := make(chan []uint32)
+	foo := func(msg []byte) {
+		data := &pb.LayerIdsResp{}
+		err := proto.Unmarshal(msg, data)
+		if err != nil {
+			log.Error("could not unmarshal layer ids response")
+		}
+		ch <- data.Ids
+	}
+
+	return ch, s.p.SendAsyncRequest(LAYER_IDS, payload, peer, foo)
+}
+
+func (s *Syncer) BlockRequestHandler(msg []byte) []byte {
+	log.Debug("handle block request")
+	req := &pb.FetchBlockReq{}
+	err := proto.Unmarshal(msg, req)
+	if err != nil {
+		return nil
+	}
+
+	block, err := s.layers.GetBlock(mesh.BlockID(req.Id))
+	if err != nil {
+		log.Error("Error handling Block request message, err:", err) //todo describe err
+		return nil
+	}
+
+	payload, err := proto.Marshal(&pb.FetchBlockResp{Id: block.Id(), Block: &pb.Block{Id: block.Id(), Layer: block.Layer()}})
+	if err != nil {
+		log.Error("Error handling request message, err:", err) //todo describe err
+		return nil
+	}
+
+	log.Debug("return block ", block)
+
+	return payload
+}
+
+func (s *Syncer) LayerHashRequestHandler(msg []byte) []byte {
+	req := &pb.LayerHashReq{}
+	err := proto.Unmarshal(msg, req)
+	if err != nil {
+		return nil
+	}
+
+	layer, err := s.layers.GetLayer(int(req.Layer))
+	if err != nil {
+		log.Error("Error handling layer request message, err:", err) //todo describe err
+		return nil
+	}
+
+	payload, err := proto.Marshal(&pb.LayerHashResp{Hash: layer.Hash()})
+	if err != nil {
+		log.Error("Error handling request message, err:", err) //todo describe err
+		return nil
+	}
+
+	return payload
+}
+
+func (s *Syncer) LayerIdsRequestHandler(msg []byte) []byte {
+	req := &pb.LayerIdsReq{}
+	err := proto.Unmarshal(msg, req)
+	if err != nil {
+		return nil
+	}
+
+	layer, err := s.layers.GetLayer(int(req.Layer))
+	if err != nil {
+		log.Debug("Error handling layer ids request message, err:", err) //todo describe err
+		return nil
+	}
+
+	blocks := layer.Blocks()
+
+	ids := make([]uint32, 0, len(blocks))
+
+	for _, b := range blocks {
```

Comment:
since you know the size of the array you don't have to use append.
it is cheaper to do:
```suggestion
	for i, b := range blocks {
	    ids[i] = b
	}
```